ephios/extra/auth.py

```diff
@@ -18,3 +37,32 @@ def update_user(self, user, claims):
         user.save()
 
         return user
+
+    def authenticate(self, request, **kwargs):
+        try:
+            self.client = EphiosOIDCClient.objects.get(id=request.session["oidc_client_id"])
+            oauth = OAuth2Session(
+                client=WebApplicationClient(client_id=self.client.client_id),
+                redirect_uri=urljoin(settings.GET_SITE_URL(), reverse("core:oauth_callback")),
+            )
+            token = oauth.fetch_token(
+                self.client.token_endpoint,
+                code=request.GET["code"],
+                client_secret=self.client.client_secret,
+                include_client_id=True,
+            )
+            self.decode_jwt_token(
+                token["id_token"]
+            )  # this already contains the claims for the tested OP, check the standard to see if we can omit the call to the user endpoint
+            user_info = oauth.request("GET", self.client.user_endpoint).json()
+            if "email" not in user_info:
+                raise SuspiciousOperation("OIDC client did not return email address")
+            users = get_user_model().objects.filter(email__iexact=user_info["email"])
+            if len(users) == 1:
+                return self.update_user(users.first(), user_info)
+            elif len(users) > 1:
+                raise SuspiciousOperation("Multiple users with same email address")
+            else:
+                return self.create_user(user_info)
+        except (KeyError, EphiosOIDCClient.DoesNotExist, InvalidTokenError):
```

Comment:
lol

Reply:
what if `self.client.user_endpoint` is not available and http 500?

Reply:
what was the lol about? ^^

Reply:
List of exceptions all with the same except block

Reply:
would you prefer something else? those are all exceptions that we can't do anything about at that moment and result in the user not being authenticated, so it made sense for me

Reply:
No I'm fine with it. Just made me chuckle